### PR TITLE
[proposal] Machine readable API specification via OpenAPI

### PR DESCRIPTION
**Pull request by takayamaki**
This PR adds machine readable API specification via OpenAPI.

By this PR,
- We can assert API schema by rspec.
- We can generate type annotation from openapi.
- We can generate API documentation from openapi.

`@openapitools/openapi-generator-cli` is frontend of openapi-generator.
openapi-generator runs on Java, I confirm that it runs on openjdk-19-jre-headless(Ubuntu 22.04).
It is needs only when edit openapi definition.

IMO, ActiveModelSerializer is almost dead.
Based on this PR, We can replace it with [alba gem](https://github.com/okuramasafumi/alba) or Pure Ruby Object that has to_h method while keeping the schema.

open api assertion via  Committee behaves like this.
image
image

Actually, if a human writes a monolithic json, openapi-generator is not needed except for validation.
https://github.com/takayamaki/mastodon/compare/main...0599ab0c57b2d9f075e06e1b8c90eaf5f1c8f6f4#diff-40bcec3448f2776bc981e50dd95b7944df172016d878957bc498e5e38e892dd5

But the effort required to maintain it is too huge for humans.

### CODE REVIEW

File: openapi/src/components/entities/null.yml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+type: string
```
Comment (takayamaki) [inline openapi/src/components/entities/null.yml:1]: This is a workaround and utility for nullable objects.

In OpanAPI 3.0, the nullable attribute is expressed as follows
```yml
properties:
  nullable_string:
    type: string
    nullable: true
```

In other hand, OpenAPI can reuse schema via `$ref`.

However, This definnition does not work as nullable attribute.
```yml
properties:
  nullable_object:
    nullable: true
    $ref: "./object.yml"
```

In this case, this workaround are works well.
```yml
properties:
  nullable_objec:
    oneOf:
      - type: string
        nullable: true
        enum:
          - null
      - $ref: "./object.yml"
```

File: tsconfig.json [hunk 2]
```diff
@@ -18,6 +18,7 @@
   "include": [
     "app/javascript/mastodon",
     "app/javascript/packs",
-    "app/javascript/types"
+    "app/javascript/types",
+    "oepnapi/lib"
```
Comment (rinsuki) [inline tsconfig.json:22]:
```suggestion
    "openapi/lib"
```

**Comment by okuramasafumi** [comment]
Hi, I'm a maintainer of Alba gem and happy to help!

**Comment by shleeable** [comment]
Java?

**Comment by renchap** [comment]
We discussed this briefly and think that having generated API clients is a good idea overall, but this is a big project and we want to consider multiple alternatives here.

Thanks for you proposal, we will have a look at it but it will most probably not be in the coming weeks, are we want to release 4.2.0 first.

**Comment by takayamaki** [comment]
I see.

The method of this PR works well in my work now.
However, I know that using openapi-generator with Java is big change.
We can consider another way.

Committee gem is a useful for Rails with openapi.
In this PR, I only use assertions, but it can also validate request parameters in production as rack middleware.
However, it can not read sepalated (uses $ref directive) specifaction.

If we find out an pure npm tool that compose uni-file openapi json from sepalated yml or json, also generator of type annoration, we may use it.

**Comment by takayamaki** [comment]
Anyway, I will keep up rebased this PR until discussion is finish.

File: spec/controllers/well_known/nodeinfo_controller_spec.rb [hunk 3]
```diff
@@ -1,41 +0,0 @@
-# frozen_string_literal: true
```
Comment (mjankowski) [inline spec/controllers/well_known/nodeinfo_controller_spec.rb:1]: Is this delete intentional?

File: app/javascript/mastodon/components/avatar.tsx [hunk 4]
```diff
@@ -1,7 +1,7 @@
 import classNames from 'classnames';
 
 import { useHovering } from '../../hooks/useHovering';
-import type { Account } from '../../types/resources';
+import type { Account } from '../../types/account';
```
Comment (mjankowski) [inline app/javascript/mastodon/components/avatar.tsx:4]: some of these changes have been done (or variations on them) in main -- rebase might remove some

File: spec/controllers/well_known/nodeinfo_controller_spec.rb (hunk 3 above)
Comment (mjankowski) [reply spec/controllers/well_known/nodeinfo_controller_spec.rb:1]: Oh nevermind - I see you moved to spec/requests

**Comment by ThisIsMissEm** [comment]
I don't really like the idea of also requiring java in your toolchain when making API changes, this feels like it adds additional barriers to contribution, since java isn't necessarily the most easy thing to install (especially with the major versions).

Is there a way we could do this without Java?

**Pull request closed without merging**